base/utils/R/sensitivity.R

```diff
@@ -51,7 +51,10 @@ read.sa.output <- function(traits, quantiles, pecandir, outdir, pft.name = "",
       run.id <- sa.run.ids[[pft.name]][quantile, trait]
       
       for(var in seq_along(variables)){
-        out.tmp <- read.output(run.id, file.path(outdir, run.id), start.year, end.year, variables[var])
+        out.tmp <- read.output(runid = run.id, outdir = file.path(outdir, run.id), 
+                               start.year = start.year, end.year = end.year, 
+                               variables = variables[var], 
+                               pft.name = pft.name)
```

Comment:
This change causes a subtle, but non-trivial, change in functionality. For example, if I previously ran a multi-PFT SA for NPP, pecan would have extracted the sensitivity of TOTAL NPP to the pft's parameters one at a time. Now we're going to get the PFT-SPECIFIC NPP's sensitivity to the pft's parameters. There doesn't appear to be any backward compatibility to get to the original functionality.  

Maybe we could change the way we SA/UA/EA variables to allow dimension specifications on variables? e.g. a SA on NPP would do total NPP while a SA on NPP[early_hardwood] would run a SA on that specific PFT?

Reply:
Any change of functionality at this basic level requires an update to the SA documentation.

Reply:
as long as there is no pft dimension in the netcdf file (we almost never use pft as a dimension in `model2netcdf`s if I'm not mistaken) read.output won't read the outputs by PFT (it will never get into the `if("pft.numbers" %in% names(nc$dim))` block)

but I see that, that check is not going to be enough because we can have both T and E output files in ED (and have pft specific outputs in LPJ-GUESS in addition to cumulative outputs) and we should be able to request an SA on NPP and DBH on the outputs of the same set of runs, then the ncdf will have a pft dimension and extract the wrong thing for NPP. 

So, we can either use variable names to specify this as you suggest or we can check the variable dimensions in the netcdf (as opposed to the broad check whether a pft dimension is specified or not in the ncdf). e.g. NPP variable dimensions will be [lat,lon,t], we'll read the whole thing, whereas DBH will have pft in its dimension, we can check  if that's the case, and only read output for that pft

Reply:
So the block in `read.output` is changed from `if("pft.numbers" %in% names(nc$dim))` to `if("pft" %in% sapply(nc$var[[v]]$dim, `[[`, "name"))` so it won't try to read pft specific SA outputs for NPP because it won't have `pft` as a dimension for models like ED2

but models like LPJ-GUESS (and I guess, Linkages as well?) can give outputs noth in total NPP and NPP per PFT. As we don't have a "NPP per PFT" variable (so that we cannot write NPP and NPP-per-PFT in our standard netcdfs), it makes sense to write NPP variable per pft for a model by default when possible, and read.output would read it per pft by default (when the variable has  `pft` as a dimension). Then the SA/EA code should know whether we want total NPP or pft specific NPP, as you suggest @mdietze:

>Maybe we could change the way we SA/UA/EA variables to allow dimension specifications on variables? e.g. a SA on NPP would do total NPP while a SA on NPP[early_hardwood] would run a SA on that specific PFT?

So it should be a tag in the xml right? e.g.

```
<sensitivity.analysis>
  <quantiles>
    <sigma>-1</sigma>
    <sigma>0</sigma>
    <sigma>1</sigma>
  </quantiles>
  <variable>DBH</variable>
  <perpft>TRUE</perpft>
</sensitivity.analysis>
```
When it doesn't see `<perpft></perpft>` tag it would default to original functionality. What do you think?

Also, does it make sense to do pft-specific ensemble? I guess we want that functionality in as well?